# Clustering Neighbourhoods in Toronto
## Coursera Capstone Project 

### Part 1 - importing libraries 

In [1]:
from bs4 import BeautifulSoup as bsoup
from urllib.request import urlopen as uReq
import requests
import lxml
import pandas as pd
from pandas import DataFrame
import numpy as np

### Reading the webpage to get the postal codes

In [2]:
my_url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M' 

In [3]:
r=requests.get(my_url)

### Parsing the web html file with BeautifulSoup package

In [4]:
# Parse the htlm with Soup
page=bsoup(r.text,"html.parser")
page


<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"XptJpgpAICsAAFpdq2oAAAAK","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":951325562,"wgRevisionId":951325562,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toronto","Ontario-related lists"]

### 'The htlm 'page'  indicates that there is a 'table class' - let's find that table 

In [5]:
rtable=page.table
rtable

<table class="wikitable sortable">
<tbody><tr>
<th>Postal code
</th>
<th>Borough
</th>
<th>Neighborhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park / Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor / Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park / Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern / Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3B
</td>
<td>North York
</td>
<td>Don Mills
</td></tr>
<tr>
<td>M4B
</td>
<td>Ea

In [6]:
results=rtable.find_all('tr')
nrows=len(results)
nrows

181

###  Such table (named 'results' contains 181 rows. The first tow is the table header.
### Therefore, there are 180 rows of data ( index 0 to 178)


In [7]:
results[0:5]

[<tr>
 <th>Postal code
 </th>
 <th>Borough
 </th>
 <th>Neighborhood
 </th></tr>,
 <tr>
 <td>M1A
 </td>
 <td>Not assigned
 </td>
 <td>
 </td></tr>,
 <tr>
 <td>M2A
 </td>
 <td>Not assigned
 </td>
 <td>
 </td></tr>,
 <tr>
 <td>M3A
 </td>
 <td>North York
 </td>
 <td>Parkwoods
 </td></tr>,
 <tr>
 <td>M4A
 </td>
 <td>North York
 </td>
 <td>Victoria Village
 </td></tr>]

### The Header of the DataFrame is the first row of data [index 0]

In [39]:
header=results[0].text.split('\n')
header

['', 'Postal code', '', 'Borough', '', 'Neighborhood', '']

### Let's check some  rows in order to prepare to build the loop that will extract all cells into a DataFrame.  For example, let's examine row 85


In [40]:
results[86].text

'\nM5M\n\nNorth York\n\nBedford Park / Lawrence Manor East\n'

In [41]:
results[86].text.split('\n')

['', 'M5M', '', 'North York', '', 'Bedford Park / Lawrence Manor East', '']

In [42]:
results[86].text.split('\n')

['', 'M5M', '', 'North York', '', 'Bedford Park / Lawrence Manor East', '']

In [43]:
Postcode=results[86].text.split('\n')[1]
Postcode

'M5M'

In [45]:
Borough=results[86].text.split('\n')[3]
Borough

'North York'

In [46]:
Neighborhood=results[86].text.split('\n')[5]
Neighborhood

'Bedford Park / Lawrence Manor East'

## Iteration loop to extract all cells into a dataframe df

In [47]:
# iteration loop to harvest all records

records =[]
n=1
while n < nrows :
    Postcode=results[n].text.split('\n')[1]
    Borough=results[n].text.split('\n')[3]
    Neighborhood=results[n].text.split('\n')[5]
    records.append((Postcode, Borough,Neighborhood))
    n=n+1

df=pd.DataFrame(records, columns=['PostalCode', 'Borough', 'Neighbourhood'])
df.head(5)



PostalCode           Borough               Neighbourhood
0        M1A      Not assigned                            
1        M2A      Not assigned                            
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

### Initial Dataframe has 180 cells or rows of data, as expected. Great!

In [48]:
df.shape

(180, 3)

### Let's check the tail of dataframe to assure all data is included as in Wikipedia table

In [49]:
df.tail()    #  Perfect! all data is included.

PostalCode       Borough  \
175        M5Z  Not assigned   
176        M6Z  Not assigned   
177        M7Z  Not assigned   
178        M8Z     Etobicoke   
179        M9Z  Not assigned   

                                         Neighbourhood  
175                                                     
176                                                     
177                                                     
178  Mimico NW / The Queensway West / South of Bloo...  
179

## Now, let's remove all rows with Borough = 'Not assigned' strings
### There are 77 Boroughs with 'Not assigned' strings



In [51]:
# How many rows have Borough equal to 'Not assigned'?
df[df['Borough']=='Not assigned'].count()

PostalCode       77
Borough          77
Neighbourhood    77
dtype: int64

In [52]:
# drops those rows where 'Not assigned' appears in column '[Borough]'
df1=df[~df.Borough.str.contains("Not assigned")]
df1=df1.reset_index(drop=True)

In [53]:
df1.head(5)

PostalCode           Borough                                 Neighbourhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [54]:
df1.tail()

PostalCode           Borough  \
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                         Neighbourhood  
98    The Kingsway / Montgomery Road  / Old Mill North  
99                                Church and Wellesley  
100              Business reply mail Processing CentrE  
101  Old Mill South / King's Mill Park / Sunnylea /...  
102  Mimico NW / The Queensway West / South of Bloo...

### The dataframe now has 103 rows (0-102) after droppong all Borough with 'Not assigned'

In [55]:
df1.shape

(103, 3)

### Now,  let's replace  'Not assigned' neighborhoods with the name of the Borough


In [56]:
df1.loc[df1['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df1['Borough']

In [57]:
df1.head()           # Now, there are no rows with 'Not assigned' strings anymore

PostalCode           Borough                                 Neighbourhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [58]:
df1.shape

(103, 3)

## Now , we will consolidate the dataframe to the unique Postal Codes.
### As per instructions, we will look for repeated postalcodes and consolidate into one postalcode with the name of the neighbourhoods merged.


### Let's examine de dataframe first to set up the best loop algorithm. How many unique postalcodes , Boroughs and Neighbourhood names are there?



In [59]:
postalcodes = df1['PostalCode'].nunique()
boroughs = df1['Borough'].nunique()
neighbourhoods= df1['Neighbourhood'].nunique()
print('Unique Postalcodes : ' + str(postalcodes))
print('Unique Boroughs  : '+ str(boroughs))
print('Unique Neighbourhoods  :' + str(neighbourhoods))

     

Unique Postalcodes : 103
Unique Boroughs  : 10
Unique Neighbourhoods  :98


### There are 103 Unique Postcodes and 10 Unique Borough with 98 Neighbourhoods

In [60]:
#df1.groupby(['PostalCode','Borough','Neighbourhood']).size().reset_index(name='Count').head()

## Let's Consolidate the dataframe to each unique PostalCodes and aggregated Neighbourhoods

In [61]:
# Starting DataFrame df1
df1.head()

PostalCode           Borough                                 Neighbourhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [62]:
nrows1=len(df1)
nrows1

103

### Let's keep df1 intact and let's work with df2 as precaution

In [63]:
# DataFrame df2 also has 103 rows or cells
df2=df1
df2.head()

PostalCode           Borough                                 Neighbourhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [64]:
# df2.index

In [65]:
# I will use two pointers, 'n and m' to sweep through the dataframe. So 'n' will vary from
#  0 to 211, and 'm ' will vary from 1 to 212
nrows2=len(df2)-1
nrows2

102

## Loop Algorithm to extract the consolidate dataframe df2 with unique postalcodes.
### We expect a df2 with 103 rows because that is the number of unique postalcodes

In [66]:
n=0

while n < nrows2 :
    post1=df2.iloc[n,0]
    #post1
    m=n+1
    post2=df2.iloc[m,0]
    #post2
    neigh1=df2.iloc[n,2]
    neigh2=df2.iloc[m,2]
    if post1==post2:
        df2.Neighbourhood[n,2] = neigh1=neigh1+','+neigh2
        #df2 = df2[df2.Neighbourhood != 'neigh2']
        df2=df2.drop(df2.index[m])
        nrows2=nrows2-1
        df2 = df2.reset_index(drop=True)
    else:
        n=n+1


#df2 = df2.reset_index(drop=True)
# print(post1, post2, 'Nigh1 is...'+ neigh1,',,,,,,,','Nigh2 is...' + neigh2, n,m)
df2.index


RangeIndex(start=0, stop=103, step=1)

In [67]:
df2.head()

PostalCode           Borough                                 Neighbourhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [68]:
df2.index

RangeIndex(start=0, stop=103, step=1)

## DataFrame consolidated to unique PostalCodes  and sorted by PostalCode 
###  103 rows by 3 columns

In [69]:
#df2.sort_values('PostalCode').head(15)
df2.head(15)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                        Neighbourhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                          Regent Park / Harbourfront  
3                   Lawrence Manor / Lawrence Heights  
4        Queen's Park / Ontario Provincial Government  
5                                    Islington Avenue  
6                                     Malvern / Rouge  
7                                           Don Mills  
8                    Parkview Hill / Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park / Princess Gardens / Martin Gr...  
12           Rouge Hill / Port Union / Highland Creek  
13                                          Don Mills  
14                                   Woodbine Heights

In [70]:
df2.shape

(103, 3)

### End of Part 1

### PART 2: Now, let's get the Latitude and Longitude coordenates for every PostalCode and add those columns to the dataframe

In [75]:

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

#import geocoder

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\joyma\anaconda3

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          92 KB

The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.50-py_0
  geopy              conda-forge/noarch::geopy-1.21.0-py_0




geographiclib-1.50   | 34 KB     |            |   0% 
geographiclib-1.50   | 34 KB     | ####7      |  47% 
geographiclib-1.50   | 34 KB     | ########## | 100% 

geopy-1.21.0         | 58 KB     |            |   0% 
geopy-1.21.0         | 58 K

###  Read the cvs file and convert it to a dataframe

In [76]:
url='http://cocl.us/Geospatial_data'
df_pcodes=pd.read_csv(url)
df_pcodes.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [77]:
df_pcodes.columns = ['PostalCode', 'Latitude', 'Longitude']
df_pcodes.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

### Now, let's sort our db2 dataframe by PostalCode, then we merge it with the df_pcodes created from the cvs file, based on common column of PostalCodes

In [78]:
df2s=df2.sort_values('PostalCode')
df2s.head()

PostalCode      Borough                             Neighbourhood
6         M1B  Scarborough                           Malvern / Rouge
12        M1C  Scarborough  Rouge Hill / Port Union / Highland Creek
18        M1E  Scarborough       Guildwood / Morningside / West Hill
22        M1G  Scarborough                                    Woburn
26        M1H  Scarborough                                 Cedarbrae

In [79]:

neighborhoods=pd.merge(df2s,df_pcodes, how='right', on = 'PostalCode')
neighborhoods.head(15)

PostalCode      Borough                                      Neighbourhood  \
0         M1B  Scarborough                                    Malvern / Rouge   
1         M1C  Scarborough           Rouge Hill / Port Union / Highland Creek   
2         M1E  Scarborough                Guildwood / Morningside / West Hill   
3         M1G  Scarborough                                             Woburn   
4         M1H  Scarborough                                          Cedarbrae   
5         M1J  Scarborough                                Scarborough Village   
6         M1K  Scarborough      Kennedy Park / Ionview / East Birchmount Park   
7         M1L  Scarborough                  Golden Mile / Clairlea / Oakridge   
8         M1M  Scarborough  Cliffside / Cliffcrest / Scarborough Village West   
9         M1N  Scarborough                       Birch Cliff / Cliffside West   
10        M1P  Scarborough  Dorset Park / Wexford Heights / Scarborough To...   
11        M1R  Scarborough                                 Wexford / Maryvale   
12        M1S  Scarborough                                          Agincourt   
13        M1T  Scarborough          Clarks Corners / Tam O'Shanter / Sullivan   
14        M1V  Scarborough  Milliken / Agincourt North / Steeles East / L'...   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577

In [80]:
neighborhoods.shape

(103, 5)

### END OF PART 2

### import libraries for part 3

In [83]:

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files


from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans


import folium # map rendering library

print('Libraries imported.')

Libraries imported.


### with the word Toronto

In [84]:
toronto_data= neighborhoods[neighborhoods['Borough'].str.contains('Toronto', na = False)].reset_index(drop=True)
toronto_data.head()

PostalCode          Borough                    Neighbourhood   Latitude  \
0        M4E     East Toronto                      The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West / Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar / The Beaches West  43.668999   
3        M4M     East Toronto                  Studio District  43.659526   
4        M4N  Central Toronto                    Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [85]:
toronto_data.shape

(39, 5)

In [86]:
latitude = 43.6532
longitude= -79.3832

In [87]:
# create map of TORONTO using latitude and longitude values above:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [93]:
CLIENT_ID='PRQJCE31TLQWGEAG4WBP0SKOG5GOLVZ4G1SVFT4PJEM4DBMV'
CLIENT_SECRET='WA1GHE1IRMZTJXTEMFFP3VP0UNBPOWG2POWICVBYYWJGQE1J'
VERSION='20200428'

In [94]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

In [95]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [96]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West / Riverdale
India Bazaar / The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park / Summerhill East
Summerhill West / Rathnelly / South Hill / Forest Hill SE / Deer Park
Rosedale
St. James Town / Cabbagetown
Church and Wellesley
Regent Park / Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond / Adelaide / King
Harbourfront East / Union Station / Toronto Islands
Toronto Dominion Centre / Design Exchange
Commerce Court / Victoria Hotel
Roselawn
Forest Hill North & West
The Annex / North Midtown / Yorkville
University of Toronto / Harbord
Kensington Market / Chinatown / Grange Park
CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst Quay / South Niagara / Island airport
Stn A PO Boxes
First Canadian Place / Underground city
Christie
Dufferin / Dovercourt Village
Little Portugal / Trinity
Brockton / Parkdale Village / Exhibition Place
High Park / 

In [97]:
print(toronto_venues.shape)
toronto_venues.head()

(1623, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                   Glen Stewart Park       43.675278       -79.294647   
4                       Upper Beaches       43.680563       -79.292869   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3               Park  
4       Neighborhood

In [98]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
Brockton / Parkdale Village / Exhibition Place                         24   
Business reply mail Processing CentrE                                  17   
CN Tower / King and Spadina / Railway Lands / H...                     16   
Central Bay Street                                                     64   
Christie                                                               17   
Church and Wellesley                                                   75   
Commerce Court / Victoria Hotel                                       100   
Davisville                                                             35   
Davisville North                                                        8   
Dufferin / Dovercourt Village                                          15   
First Canadian Place / Underground city                               100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East / Union Station / Toronto Isl...                    100   
High Park / The Junction South                                         22   
India Bazaar / The Beaches West                                        23   
Kensington Market / Chinatown / Grange Park                            57   
Lawrence Park                                                           3   
Little Portugal / Trinity                                              41   
Moore Park / Summerhill East                                            1   
North Toronto West                                                     22   
Parkdale / Roncesvalles                                                13   
Queen's Park / Ontario Provincial Government                           40   
Regent Park / Harbourfront                                             48   
Richmond / Adelaide / King                                             95   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede / Swansea                                                    39   
St. James Town                                                         81   
St. James Town / Cabbagetown                                           42   
Stn A PO Boxes                                                         97   
Studio District                                                        40   
Summerhill West / Rathnelly / South Hill / Fore...                     16   
The Annex / North Midtown / Yorkville                                  23   
The Beaches                                                             5   
The Danforth West / Riverdale                                          43   
Toronto Dominion Centre / Design Exchange                             100   
University of Toronto / Harbord                                        34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
Brockton / Parkdale Village / Exhibition Place                          24   
Business reply mail Processing CentrE                                   17   
CN Tower / King and Spadina / Railway Lands / H...                      16   
Central Bay Street                                                      64   
Christie                                                                17   
Church and Wellesley                                                    75   
Commerce Court / Victoria Hotel                                        100   
Davisville                                                              35   
Davisville North                    

### Let's find out how many unique categories can be curated from all the returned venues

In [99]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 226 uniques categories.


### Analyze Each Neighborhood

In [100]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols=list(toronto_onehot.columns.values)
cols.pop(cols.index('Neighborhood'))
toronto_onehot=toronto_onehot[['Neighborhood']+cols]

# rename Neighborhood for Neighbourhood so that future merge works
toronto_onehot.rename(columns = {'Neighborhood': 'Neighbourhood'}, inplace = True)
toronto_onehot.head()

Neighbourhood  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   The Beaches        0                   0             0               0   
1   The Beaches        0                   0             0               0   
2   The Beaches        0                   0             0               0   
3   The Beaches        0                   0             0               0   
4   The Beaches        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  \
0           0                   0       0           0              0   
1           0                   0       0           0              0   
2           0                   0       0           0              0   
3           0                   0       0           0              0   
4           0                   0       0           0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  Butcher  \
0                0         0             0              0         0        0   
1                0         0             0              0         0        0   
2                0         0             0              0         0        0   
3                0         0             0              0         0        0   
4                0         0             0              0         0        0   

   Cable Car  Café  Cajun / Creole Restaurant  Candy Store  \
0          0     0                          0            0   
1          0     0                          0            0   
2          0     0                          0            0   
3          0     0                          0            0   
4          0     0                  

In [101]:

toronto_onehot.shape

(1623, 226)

In [102]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Airport  \
0                                         Berczy Park   0.0000   
1      Brockton / Parkdale Village / Exhibition Place   0.0000   
2               Business reply mail Processing CentrE   0.0000   
3   CN Tower / King and Spadina / Railway Lands / ...   0.0625   
4                                  Central Bay Street   0.0000   
5                                            Christie   0.0000   
6                                Church and Wellesley   0.0000   
7                     Commerce Court / Victoria Hotel   0.0000   
8                                          Davisville   0.0000   
9                                    Davisville North   0.0000   
10                      Dufferin / Dovercourt Village   0.0000   
11            First Canadian Place / Underground city   0.0000   
12                           Forest Hill North & West   0.0000   
13                           Garden District, Ryerson   0.0000   
14  Harbourfront East / Union Station / Toronto Is...   0.0000   
15                     High Park / The Junction South   0.0000   
16                    India Bazaar / The Beaches West   0.0000   
17        Kensington Market / Chinatown / Grange Park   0.0000   
18                                      Lawrence Park   0.0000   
19                          Little Portugal / Trinity   0.0000   
20                       Moore Park / Summerhill East   0.0000   
21                                 North Toronto West   0.0000   
22                            Parkdale / Roncesvalles   0.0000   
23       Queen's Park / Ontario Provincial Government   0.0000   
24                         Regent Park / Harbourfront   0.0000   
25                         Richmond / Adelaide / King   0.0000   
26                                           Rosedale   0.0000   
27                                           Roselawn   0.0000   
28                                Runnymede / Swansea   0.0000   
29                                     St. James Town   0.0000   
30                       St. James Town / Cabbagetown   0.0000   
31                                     Stn A PO Boxes   0.0000   
32                                    Studio District   0.0000   
33  Summerhill West / Rathnelly / South Hill / For...   0.0000   
34              The Annex / North Midtown / Yorkville   0.0000   
35                                        The Beaches   0.0000   
36                      The Danforth West / Riverdale   0.0000   
37          Toronto Dominion Centre / Design Exchange   0.0000   
38                    University of Toronto / Harbord   0.0000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0               0.0000        0.0000           0.000           0.0000   
1               0.0000        0.0000           0.000           0.0000   
2               0.0000        0.0000           0.000           0.0000   
3               0.0625        0.0625           0.125           0.1875   
4               0.0000        0.0000           0.000           0.0000   
5               0.0000        0.0000           0.000           0.0000   
6               0.0000        0.0000           0.000           0.0000   
7               0.0000        0.0000           0.000           0.0000   
8               0.0000        0.0000           0.000           0.0000   
9               0.0000        0.0000           0.000           0.0000   
10              0.0000        0.0000           0.000           0.0000   
11              0.0000        0.0000           0.000           0.0000   
12              0.0000        0.0000           0.000           0.0000   
13              0.0000        0.0000           0.000           0.0000   
14              0.0000        0.0000           0.000           0.0000   
15              0.0000        0.0000           0.000           0.0000   
16              0.0000        0.0000           0.000           0.0000   
17              0.0000        0.0000           0.000           0.0000   
18          

In [103]:
toronto_grouped.shape

(39, 226)

### Let's print each neighborhood along with the top 5 most common venues

In [105]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
          venue  freq
0   Coffee Shop  0.07
1  Cocktail Bar  0.05
2        Bakery  0.03
3      Beer Bar  0.03
4    Restaurant  0.03


----Brockton / Parkdale Village / Exhibition Place----
                   venue  freq
0                   Café  0.12
1            Coffee Shop  0.08
2         Breakfast Spot  0.08
3              Nightclub  0.08
4  Performing Arts Venue  0.04


----Business reply mail Processing CentrE----
                venue  freq
0  Light Rail Station  0.12
1         Yoga Studio  0.06
2              Garden  0.06
3                Park  0.06
4          Comic Shop  0.06


----CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst Quay / South Niagara / Island airport----
              venue  freq
0   Airport Service  0.19
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3           Airport  0.06
4               Bar  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.20
1  Italian Restaurant  0.

### First, let's write a function to sort the venues in descending order.

In [106]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Now let's create the new dataframe and display the top 10 venues for each neighborhood

In [107]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighbourhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1      Brockton / Parkdale Village / Exhibition Place                  Café   
2               Business reply mail Processing CentrE    Light Rail Station   
3   CN Tower / King and Spadina / Railway Lands / ...       Airport Service   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley      Sushi Restaurant   
7                     Commerce Court / Victoria Hotel           Coffee Shop   
8                                          Davisville        Sandwich Place   
9                                    Davisville North          Dance Studio   
10                      Dufferin / Dovercourt Village              Pharmacy   
11            First Canadian Place / Underground city           Coffee Shop   
12                           Forest Hill North & West         Jewelry Store   
13                           Garden District, Ryerson        Clothing Store   
14  Harbourfront East / Union Station / Toronto Is...           Coffee Shop   
15                     High Park / The Junction South    Mexican Restaurant   
16                    India Bazaar / The Beaches West  Fast Food Restaurant   
17        Kensington Market / Chinatown / Grange Park                  Café   
18                                      Lawrence Park                  Park   
19                          Little Portugal / Trinity                   Bar   
20                       Moore Park / Summerhill East            Playground   
21                                 North Toronto West        Clothing Store   
22                            Parkdale / Roncesvalles             Gift Shop   
23       Queen's Park / Ontario Provincial Government           Coffee Shop   
24                         Regent Park / Harbourfront           Coffee Shop   
25                         Richmond / Adelaide / King           Coffee Shop   
26                                           Rosedale                  Park   
27                                           Roselawn                Garden   
28                                Runnymede / Swansea           Coffee Shop   
29                                     St. James Town                  Café   
30                       St. James Town / Cabbagetown            Restaurant   
31                                     Stn A PO Boxes           Coffee Shop   
32                                    Studio District                  Café   
33  Summerhill West / Rathnelly / South Hill / For...           Coffee Shop   
34              The Annex / North Midtown / Yorkville        Sandwich Place   
35                                        The Beaches                  Park   
36                      The Danforth West / Riverdale      Greek Restaurant   
37          Toronto Dominion Centre / Design Exchange           Coffee Shop   
38                    University of Toronto / Harbord                  Café   

   2nd Most Common Venue  3rd Most Common Venue  \
0           Cocktail Bar                 Bakery   
1            Coffee Shop         Breakfast Spot   
2            Yoga Studio          Auto Workshop   
3         Airport Lounge       Airport Terminal   
4     Italian Restaurant                   Café   
5                   Café                   Park   
6    Japanese Restaurant            Coffee Shop   
7             Restaurant                   Café   
8           Dessert Shop            Pizza Place   
9                  Hotel                   Park   
10                Bakery             Smoke Shop   
11                  Café                  Hotel   
12                 Trail     Mexican Restaurant   
13           Coffee Shop                   Café   
14              Aquarium                  Hotel   
15                  Café        Thai Restaurant   
16          

### Run k-means to cluster the neighbourhood into 5 clusters

In [108]:

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

### Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood

In [109]:
toronto_merged = toronto_data

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                    Neighbourhood   Latitude  \
0        M4E     East Toronto                      The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West / Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar / The Beaches West  43.668999   
3        M4M     East Toronto                  Studio District  43.659526   
4        M4N  Central Toronto                    Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0                  Park                   Pub   
1 -79.352188               0      Greek Restaurant           Coffee Shop   
2 -79.315572               0  Fast Food Restaurant           Pizza Place   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               0                  Park              Bus Line   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
0                 Trail     Health Food Store                  Yoga Studio   
1    Italian Restaurant        Ice Cream Shop                    Bookstore   
2        Sandwich Place    Italian Restaurant                   Restaurant   
3             Gastropub                Bakery                      Brewery   
4           Swim School         Deli / Bodega  Eastern European Restaurant   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                   Diner         Deli / Bodega      Department Store   
1  Furniture / Home Store    Frozen Yogurt Shop        Cosmetics Shop   
2              Steakhouse        Ice Cream Shop            Board Shop   
3     American Restaurant     Convenience Store        Sandwich Place   
4     Dumpling Restaurant            Donut Shop      Doner Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0          Dessert Shop    Distribution Center  
1               Brewery        Bubble Tea Shop  
2               Brewery          Burrito Place  
3           Cheese Shop              Pet Store  
4               Dog Run    Distribution Center

## Final resulting clusters visualization - Toronto Neighborhoods

In [110]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Now, one can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, one can then assign a name to each cluster, if so desired

In [111]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               0                  Park   
1       East Toronto               0      Greek Restaurant   
2       East Toronto               0  Fast Food Restaurant   
3       East Toronto               0                  Café   
4    Central Toronto               0                  Park   
5    Central Toronto               0          Dance Studio   
6    Central Toronto               0        Clothing Store   
7    Central Toronto               0        Sandwich Place   
8    Central Toronto               0            Playground   
9    Central Toronto               0           Coffee Shop   
10  Downtown Toronto               0                  Park   
11  Downtown Toronto               0            Restaurant   
13  Downtown Toronto               0           Coffee Shop   
14  Downtown Toronto               0        Clothing Store   
15  Downtown Toronto               0                  Café   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
19  Downtown Toronto               0           Coffee Shop   
21  Downtown Toronto               0           Coffee Shop   
22   Central Toronto               0                Garden   
23   Central Toronto               0         Jewelry Store   
24   Central Toronto               0        Sandwich Place   
25  Downtown Toronto               0                  Café   
28  Downtown Toronto               0           Coffee Shop   
29  Downtown Toronto               0           Coffee Shop   
30  Downtown Toronto               0         Grocery Store   
31      West Toronto               0              Pharmacy   
32      West Toronto               0                   Bar   
33      West Toronto               0                  Café   
34      West Toronto               0    Mexican Restaurant   
36      West Toronto               0           Coffee Shop   
37  Downtown Toronto               0           Coffee Shop   
38      East Toronto               0    Light Rail Station   

   2nd Most Common Venue  3rd Most Common Venue  \
0                    Pub                  Trail   
1            Coffee Shop     Italian Restaurant   
2            Pizza Place         Sandwich Place   
3            Coffee Shop              Gastropub   
4               Bus Line            Swim School   
5                  Hotel                   Park   
6            Coffee Shop            Yoga Studio   
7           Dessert Shop            Pizza Place   
8            Yoga Studio           Dance Studio   
9                    Pub            Supermarket   
10                 Trail             Playground   
11           Coffee Shop                   Café   
13                   Pub                 Bakery   
14           Coffee Shop                   Café   
15           Coffee Shop           Cocktail Bar   
16          Cocktail Bar                 Bakery   
17    Italian Restaurant                   Café   
19              Aquarium                  Hotel   
21            Restaurant                   Café   
22           Yoga Studio           Dance Studio   
23                 Trail     Mexican Restaurant   
24                  Café            Coffee Shop   
25            Restaurant    Japanese Restaurant   
28    Italian Restaurant                   Café   
29                  Café                  Hotel   
30                  Café                   Park   
31                Bakery             Smoke Shop   
32            Restaurant  Vietnamese Restaurant   
33           Coffee Shop         Breakfast Spot   
34                  Café        Thai Restaurant   
36           Pizza Place                   Café   
37      Sushi Restaurant                  Diner   
38           Yoga Studio          Auto Workshop   

            4th Most Common Venue        5th Most Common Venue  \
0               Health Food Store                  Yoga Studio   
1                  Ice Cream Shop    

# End of part 3